lsp/source/regenerateTextmate.civet

```diff
@@ -0,0 +1,174 @@
+* as vscode from 'vscode'
+* as path from 'node:path'
+{ promises as fsPromises } from 'node:fs'
+
+{ findConfig, loadConfig as loadCivetConfig } from '@danielx/civet/config'
+{ features } from './grammarList.civet'
+
+let output: vscode.OutputChannel | undefined
+
+// Entry point for VS Code command
+export async function regenerateTextmate(showOutput = false)
+  if not output
+    output = vscode.window.createOutputChannel("Civet Textmate Grammar")
+  else
+    output!.clear()
+  out := output!
+  if showOutput
+    out.show()
+  out.appendLine("Starting Civet grammar regeneration...")
+
+  civetParseOpts .= {}
+  optionsModified .= false
+
+  try
+    workspaceFolder := vscode.workspace.workspaceFolders?.[0]
+    if not workspaceFolder
+      msg := 'Civet: No workspace folder – cannot regenerate grammar'
+      out.appendLine(`ERROR: ${msg}`)
+      vscode.window.showErrorMessage(msg)
+      return { modified: false, parseOptions: {} }
+
+    // 1 Load Civet config (may return undefined)
+    configPath := workspaceFolder.uri.fsPath
+    out.appendLine(`Searching for Civet config in: ${configPath}`)
+    cfgFile := await findConfig(configPath).catch((err) => {
+      out.appendLine(`ERROR while searching for config: ${err?.message ?? err}`)
+      return undefined
+    })
+
+    if cfgFile
+      out.appendLine(`Found Civet config: ${cfgFile}`)
+    else
+      out.appendLine('No Civet config file found – using defaults')
+
+    cfg := cfgFile ? await loadCivetConfig(cfgFile).catch((err) => {
+      out.appendLine(`ERROR loading config: ${err?.message ?? err}`)
+      return undefined
+    }) : undefined
+    civetParseOpts = cfg?.parseOptions ?? {}
+    out.appendLine(`Loaded parseOptions: ${JSON.stringify(civetParseOpts, null, 2)}`)
+
+    // If parse options are empty, skip writing cache/grammar
+    if Object.keys(civetParseOpts).length is 0
+      out.appendLine('Warning: civetParseOpts is empty, skipping grammar/cache update.')
+      return { modified: false, parseOptions: civetParseOpts }
+    
+    // 2 Locate grammar file within the extension bundle
+    ext := vscode.extensions.getExtension 'DanielX.civet'
+    if !ext
+      msg := 'Civet: Unable to locate extension path'
+      out.appendLine(`ERROR: ${msg}`)
+      vscode.window.showErrorMessage(msg)
+      return { modified: false, parseOptions: civetParseOpts }
+    
+    grammarFile := path.join ext.extensionPath, 'syntaxes', 'civet.json'
+    cacheFile := path.join ext.extensionPath, 'syntaxes', '.civet-grammar-cache.json'
+    out.appendLine(`Found grammar file at: ${grammarFile}`)
+
+    // Check if options have changed by comparing with cache (async)
+    cachedOpts .= {}
+    try
+      await fsPromises.access cacheFile
```

Comment:
We shouldn't call `access` before `readFile`. There's no guarantee that the file remains in the same state after the call to access and readFile will error anyway if the file is unreadable.

https://nodejs.org/api/fs.html#fspromisesaccesspath-mode